### DISCUSSION TITLE
Making a GUI

### DISCUSSION THREAD

**Discussion by JamesPerisher** [ThioJoe/YT-Spammer-Purge#84] [Ideas]
Hey  @ThioJoe 

I am wanted to make a PR to add a GUI

From experience getting a GUI to work without breaking cross-platform support is a little bit difficult and also requires a library in windows it is possible to use Tkinter however kivy is a more cross platform solution.  I have quite a bit of experience with kivy as can be seen here https://github.com/pauln07org/Kryptos as I have developed a full end2end encrypted messaging app using kivy.

I understand that you want to keep it simple so you can understand the code and this addition would add what can be considered complexity.

But as the contributing guidelines ask to open an issue before adding a new library or complicated code, please enlighten me with your thoughts on this.

**Comment by ThioJoe** [comment]
I tried making a GUI with Qt and failed miserably. I personally thought anything with tkinter looked like 1995, and kind of gave up.

I'm actually working on re-writing the entire script in C# / .NET so I can make a compiled cross platform version, especially with the new .NET MAUI framework (though it's still in beta). I'd hold off on starting it for now. I'll keep this open though and let you know what I decide to do. If I do decide to do the GUI with python, I may very well require your help.

**Reply by TechStudent10** [reply]
@ThioJoe If you want to write the entire thing using C#, can't you use CPython to translate or something? I am not an expert on CPython, but it might be what you're looking for if you want to write the program in C or C#.

**Reply by Gowixx** [reply]
@ThioJoe there is a GUI designer for PyQt5. It works very well and the product looks amazing. Just `pip install pyqt5-tools` and run `pyqt5-tools designer` in cmd. You can also convert the `.ui` file it exports into a nice and clean `.py` file. I have used it a couple times before and to say the least, it's a very smooth and clean expirience.

**Reply by TechStudent10** [reply]
@Gowixx Well @ThioJoe could use PyQt5 or 6, but an alternative could be just writing it (the UI) in HTML, CSS and JavaScript. I remember seeing a library that allows you to make GUIs using HTML, CSS and JavaScript while having access to Python. For example, I have this function:

```python
def delete_comments():
    print("Delete comments here")
```

Well, using a class to make it a JS API:

```python
class JSAPI:
    def delete_comments():
        print("Delete comments here")
```

Then, you can access it in JS:

```javascript
window.python.delete_comments()
```

Or something like that. I forgot the module name, but when I remember, I'll edit this comment to say so.

Edit: Just found it! It's called `pywebview`. Link to documentation: https://pywebview.flowrl.com/.

**Reply by TechStudent10** [reply]
According to https://pywebview.flowrl.com/guide/api.html#webview-create-window:

> `js_api` - Expose a python object to the DOM of the current `pywebview` window. Methods of the `js_api` object can be executed from Javascript by calling `window.pywebview.api.<methodname>(<parameters>)`. Please note that the calling Javascript function receives a promise that will contain the return value of the python function. Only basic Python objects (like int, str, dict, ...) can be returned to Javascript.

So yep, this is possible.

**Reply by Gowixx** [reply]
Interesting find

**Reply by TechStudent10** [reply]
@Gowixx Thanks

**Reply by Ahsoka** [reply]
@TechStudent11 Another option for interacting with JS from within Python is [Eel](https://github.com/ChrisKnott/Eel).  One thing to consider is that using an entirely Python based option such PyQt or wxPython may be easier than learning/working with multiple languages.

**Reply by TechStudent10** [reply]
@Ahsoka I know that Eel exists. It's just I found pywebview to be easier. And learning/working with multiple languages can be a pain if the other languages are fairly complicated. HTML and CSS are simple and JS has a similar syntax as Python. I easily learned JavaScript because of Python.

**Reply by rbeer** [reply]
*Electron has entered the chat*

**Reply by TechStudent10** [reply]
@rbeer oh no...


wait, how are you gonna communicate with Python and Electron?

**Reply by rbeer** [reply]
You don't. You write the whole thing in it, if you want to go with the HTML5 stack. It's out-of-the-box cross-platform, and for this use case faster.

Merging languages like you proposed is almost always a headache.

If you really want to keep the existing codebase, you make it a backend-frontend solution, adding some simple API to this here and have your frontend (and any other thereafter) query that. Apply the tried and true first rule of the [UNIX philosophy](https://en.m.wikipedia.org/wiki/Unix_philosophy#Origin):
```Make each program do one thing well. To do a new job, build afresh rather than complicate old programs by adding new "features".```

Since there seems to be an interest in turning this into a SaaS with spam detection powered by TensorFlow, I'd go with the latter . The idea of adding an item to the YouTube menu (via extension) would be another type of frontend. So you end up with:

1. Service, derived from this basis to handle deletion.
2. Service to train the text classifier
3. Service to receive spam reports from trusted sources, e.g. vetted channel owners, that get fed as training sets into (2)
4. Frontend for those neat graphs and statistics or whatever you guys need a GUI for
5. An extension for the easy reporting option.

Let this thing run until the model is sufficiently trained and offer/activate an auto-removal feature based on that model.

Half of this probably doesn't make much sense or is off-topic in this thread, but it's 5am here and I'm going to bed now. 😴

P.S. C# is nice and all, but you will get the same Linux and MacOS people asking how to install mono runtime and all that jazz.

**Reply by TechStudent10** [reply]
@rbeer You got two things right (kind of). One, writing a program in one language instead of multiple is easier, but it kind of is inevitable in certain circumstances. And two, most that didn't make sense. I actually already found someone how made an extension after watching @ThioJoe's video in JavaScript (#88). So, it is possible.

But still, the Python codebase is easy to understand. So both approaches have their advantages and disadvantages, so this @ThioJoe's decision.

**Comment by JamesPerisher** [comment]
In C# is asking for torture and I totally agree about tkinter being 1995 lol. If you where having issues doing it in python C# isn't going to be any easier trust me. I realy like the project so I'll spend an hour tomorrow most likely can get 95% complete in that time I'll open a PR just so u can look at my code.

**Comment by cob16** [comment]
In the far future would you want this to be a webpage/service?

I mention it as you could always run your GUI as a localhost webpage with plans to make it a hosted site in the future?
Not as cool as the options you have mentioned for sure but thought I might throw it out there

**Reply by TechStudent10** [reply]
See #40

**Comment by Ahsoka** [comment]
I have a lot experience using PyQt6 so if you want to go that direction I would be happy to help.  See here for reference: https://github.com/Ahsoka/beskar

**Comment by Stridsvagn69420** [comment]
I believe porting the code to C# itself isn't that hard (e.g. there's an official .NET Google API library) but making the GUI is, and you would have to wait until 2nd quarter 2022 for .NET MAUI to officially be released.

My suggestion would be to make it hosted on the web or even a PWA (thus you basically have a GUI version for literally every platform) and/or use Electron for a cross-platform desktop GUI version that's also independent from any server.
It should be easy to make as well since the language in both cases is primarily JavaScript (although I recommend TypeScript).

**Reply by Masamune3210** [reply]
The only major problem I have with Electron is it has an unfortunate habit of bloating both the file size and the performance requirements. Something slimmer would probably be better if possible

**Reply by Ahsoka** [reply]
> The only major problem I have with Electron is it has an unfortunate habit of bloating both the file size and the performance requirements. Something slimmer would probably be better if possible

@Masamune3210

What would be an alternative to Electron that is slimmer?

**Reply by Masamune3210** [reply]
I don't know off the top of my head, was more just putting in my two cents

**Reply by Stridsvagn69420** [reply]
Tauri is a good alternative to Electron. And afaik, you don't have to write *that* much Rust code. You'd only need some things for the file system and that's pretty easy. You might not even need to write Rust code at all.

**Comment by ThioJoe** [comment]
Converting to discussion

**Comment by JamesPerisher** [comment]
I had a though that I think would be a good direction to take the project. If the application where to be written/ported to javascript then it can be a quite easily made into a browser extension that would work on all platforms allowing total modification of the YouTube comment section allowing the delete all messages by userxyz to be a direct button in the youtube.com comment section.
By doing this the entire concept of a gui is nolonger needed as an item can simply be added to an existing dropdown like so:
![image](https://user-images.githubusercontent.com/37967469/142713969-833d7c58-69f8-48e0-af1f-1ae578784d0e.png)

**Reply by TechStudent10** [reply]
@JamesPerisher wait, that's actually a good idea. How come nobody thought of it!

**Reply by JamesPerisher** [reply]
because its probably better to keep it python however i think it might be possible to use Pyjamas but idk

**Comment by jac0-b** [comment]
I guess I'll throw my hat into the ring

I really think a python flask (or django) app is the best option here. Firstly this project massively lends itself to python so deviating from that sounds like a bad idea and using python also has the added benefit of lowering the barrier to contribute. Secondly a webapp also lends itself well to the project and can be run on localhost and then expanded if needed and hosted externally.

I don't think writing a desktop GUI is a good idea here. As you can see from my profile I am quite experienced with PyQt and still believe it's not worth the hassle over a flask app. Either way it's probably worth letting the project mature some more before committing to a UI as it's constantly changing at the moment (a ui test branch could be useful though).

**Reply by TechStudent10** [reply]
Agreed. Leaving the app in Python is how it should be.

**Reply by markio00** [reply]
i was about to suggest the same, i feel like otherwise it will lead to some sort of fragmentation and that adds complexity reducing maintainability.
And on top of that it's crazy simple to setup

**Comment by Kenshin9977** [comment]
PySimpleGUI is the easiest GUI framework for Python imo. It took me one hour to start with it

**Reply by sbordeyne** [reply]
PySimpleGUI though is hell to use when you're not doing just a basic form. It's not a framework that I'd recommend except for micro apps. Furthermore it uses tkinter under the hood, there's no benefit to that over just tkinter, and it's a detrimental dependancy as the project scales up.

Given the amount of filtering options available (with more coming in the future), it would be a poor choice in my opinion.

**Reply by DataStream3** [reply]
> 

Eh, not really. PySimpleGUI is not that hard to use. I've been using it for a decently sized project (one with multiple windows) and it's pretty flexible. Wasn't easy to learn but once you do it's not bad. Very well documented as well. 

I could try making a GUI for this project using it.

**Comment by Void48** [comment]
GUI would be really nice 👍🏻

**Comment by FranMaric** [comment]
Hey, I just found out about this project and my first thought was that it needs a GUI. I am experienced with Flutter so I might be biased.  I see that you don't want new dependencies and I understand that but Flutter might be a good option because you can run it on web if you want to host it or run it in desktop environment without any issues (very portable). How difficult would it be for the python scripts to talk with dart (Flutter code)? If possible at all?

**Reply by Stridsvagn69420** [reply]
If using the existing Python code, you'd need Python already installed (only works on desktop) or some Python runtime/interpreter for Flutter/Dart, and I think this will get complicated. It might work though.

Another option would be to port the Python code to Dart, which also would be better for package size and speed (there are no external dependencies too). There's also a [Dart version of the Googleapis](https://github.com/google/googleapis.dart), so my best guess is that you can just copy the existing Python code to Dart.

I personally prefer the latter. I know this means that the entire code needs to be ported to Dart, which could take some time but would probably be worth it.

Btw, I don't know much about Flutter but some people might be using Spammer Purge only in the console. Is there an option to not run the GUI, e.g. with a command-line argument, so that both console and GUI versions are supported and only have one code base?

**Reply by FranMaric** [reply]
I think it would benefit from being rewritten in dart but it seems like a lot of work. Dart could just send commands to console and talk to python code through console. That would allow to have GUI and command-line interface.

**Reply by Stridsvagn69420** [reply]
I just found this:
https://github.com/srplab/starcore_for_flutter
Apparently it allows running Python code from Flutter, it seems a bit complex though.
I have no experience with it, but it might help.

**Comment by Gowixx** [comment]
These all seem like great idea, however, I think that using flask would be the best idea. No need to work with extra dependencies, just run  the flask server, and its easy. Using flask is also easy, as long as you learn the basic syntax.

**Reply by TechStudent10** [reply]
Agreed. I really like using Flask for building websites that require backend code.

**Reply by Gowixx** [reply]
Or, alternatively, PyQt5 could be used with the QtDesigner

**Reply by TechStudent10** [reply]
> Or, alternatively, PyQt5 could be used with the QtDesigner

Also a nice idea

**Comment by chekoduadarsh** [comment]
Hey, just found this project, great work guys!! 

I agree with @Gowixx flask is simple and perfect for this tool. But I think this tool will have more out reach if it is a web extension. Something like AdBlockPlus  "SpamPurgePlus". This will reduce the time spent on purging spams and will be easy to use.

**Reply by Firecul** [reply]
I do like the idea of an extension even more than a GUI. You are already in the web browser so ideally, you'd only have to click 1 or two times to run it.

**Reply by ThioJoe** [reply]
To be honest I'm not sure how it would really make sense as a chrome extension, or what advantage there would be to have it as one.

**Reply by Firecul** [reply]
> what advantage there would be to have it as one

Ease of use and wouldn't require an external program. 
Potentially 1 click rather than launching the program, entering your URL, choosing scan mode etc

**Reply by Ahsoka** [reply]
I'm not a fan of an extension either. I feel like an extension doesn't give enough real-estate to properly implement all the controls and configuration of the app, without having a lot sub menus or making everything feel super cramped. Additionally, extensions are not universal so you would have to create a separate one for each browser. And on top of that extensions are exclusively in JavaScript (at least for Chrome) so this means the codebase would have to be entirely rewritten.

**Reply by chekoduadarsh** [reply]
> To be honest I'm not sure how it would really make sense as a chrome extension, or what advantage there would be to have it as one.

1. People are already familiar with ABP so the learning difference will be short.
2. You can drop hassle installation process and it can run Cross OS !!
3. as @Firecul  mentioned : Reducing the manual effort to report spams ("Increasing the number of reports")

**Reply by ThioJoe** [reply]
This is absolutely nothing like adblock plus though

**Reply by chekoduadarsh** [reply]
@Ahsoka Ya that's true!! its pain rewrite the code-base. For GUI cramping we can do some UX engineering and see. But I get your point. 

What problem I was trying to address here is to make it easy to report spams. That is the only way AI can learn. So more people do it the netter it will be.

As a middle ground we can make a fully functional web APP, and a small extension with minimal capabilities(which can open the web app)

**Reply by chekoduadarsh** [reply]
> This is absolutely nothing like adblock plus though

Ya ik sry, Since I thought ultimate functionality (Removing unwanted stuff) is similar i just pulled it as an example!!

**Reply by Ahsoka** [reply]
> @Ahsoka Ya that's true!! its pain rewrite the code-base. For GUI cramping we can do some UX engineering and see. But I get your point. 
> 
> 
> 
> What problem I was trying to address here is to make it easy to report spams. That is the only way AI can learn. So more people do it the netter it will be.
> 
> 
> 
> As a middle ground we can make a fully functional web APP, and a small extension with minimal capabilities(which can open the web app)
> 
> 

I think if you really want to make it accessible via the browser the best option would be to write it as a website, since a website (for the most part) doesn't have to be rewritten for each separate browser, however, as @ThioJoe has stated before he does not want to have the responsibility of storing user authentication credentials, due to the possibility of a data breach. Perhaps a website could still be used by storing credentials on the client side via cookies or the browser's local storage.

**Reply by TechStudent10** [reply]
A suggestion might be that the user uploads their `client_secrets.json` file to a website. The website then encodes/encrypts the file so that attackers can't use it (although I'm not sure how having access to the file would warrant a potential security risk. Maybe it allows you to go on a user's behalf if they've already authenticated?). It will then use the file so that @ThioJoe doesn't have to worry about credentials. Once the user has left, the program will automatically delete the file, wiping from the server.

Sounds complicated, not even entirely sure it's possible, but hey it's an idea.

**Reply by Ahsoka** [reply]
> A suggestion might be that the user uploads their `client_secrets.json` file to a website. The website then encodes/encrypts the file so that attackers can't use it (although I'm not sure how having access to the file would warrant a potential security risk. Maybe it allows you to go on a user's behalf if they've already authenticated?). It will then use the file so that @ThioJoe doesn't have to worry about credentials. Once the user has left, the program will automatically delete the file, wiping from the server.
> 
> Sounds complicated, not even entirely sure it's possible, but hey it's an idea.

Any client-server model always comes with risk.  The server will always be the single point of failure and no matter if the client stores the credentials on their computer at some point the server will need to read in these credentials. If an attacker gains control of the server they could manipulate the process in which the client sends the credentials to server to do malicious things for example, forward the credentials to themselves.  Risk can be mitigated by letting reputable services host the website for you so you don't have to really worry securing your servers (all you need to do is upload code and the hosting service does the rest).  However, like I mentioned earlier, websites always come with risk.  It is really hard to tell how much of lucrative target this app would be for any potential attackers, but if popular creators use this it might be worth an attackers time to hack this service and use it to take over an account and demand ransom for the account.  Of course the situation I just mentioned is unlikely but possible.  It's ultimately up to @ThioJoe to determine how much risk he is willing to take on.  In the current implementation there is very little risk and the only attack vector I can think of is if someone compromised the updating portion of the app and uploaded malicious code.

**Reply by Firecul** [reply]
I'd vote against the website idea, it would possibly be simpler but I don't think this is a good fit for a website.

**Reply by ThioJoe** [reply]
That's one of the big reasons I haven't pursued the hosted version already - I don't want to be responsible for the security of hundreds thousands of login credentials. The security would be too easy to mess up, and I literally have zero experience with any of that so I wouldn't trust ability to secure it well enough. 

If there was going to be a web UI, if anything it would be local only, I think that's what some have been suggesting anyway.

**Reply by TechStudent10** [reply]
@ThioJoe If that's route you want to go, may I suggest It's called [`pywebview`](https://pywebview.flowrl.com/)? It allows you to create a window with a webview that allows you to render HTML code. It's written in Python and it doesn't require a full rewrite.

I've only done small tests with it, but I do think it is a good option.

**Comment by Firecul** [comment]
> extensions are not universal so you would have to create a separate one for each browser. And on top of that extensions are exclusively in JavaScript (at least for Chrome) so this means the codebase would have to be entirely rewritten.

This is the biggest flaw of the extension idea I think, having to rewrite from scratch in a different language, possibly more than once.

> a small extension with minimal capabilities

Basically, this is what I was thinking: 
 - An extension with most of the functionality of YTSpammerPurge but slightly streamlined. 
 - When you are on a YT video a button could show up next to the like/dislike or (un)subscribe buttons, when you click that it would run an auto scan with the least controversial rules and automatically reports found items (assuming low false-positive %)
 - When you are on a channel page a button next to (un)subscribe would scan the last 5 videos of that creator with the same scan as above.
 - For creators, when on your dashboard page there would be a button to scan your channel or a number of videos, configurable in the extension settings.

More could be available from a dropdown button in the extensions bar but basic functionality would only require it for instructions creating and adding the API key

**Reply by Ahsoka** [reply]
In my other posts I have made my stance clear but that does not mean that I am not open to people pursuing other avenues or implementations.  I would love to see some potential mockups of your idea if possible.  Maybe some drawings on a screenshot or perhaps if you have the time some basic graphics, kind of like what was done [in this post](https://github.com/ThioJoe/YT-Spammer-Purge/discussions/84#discussioncomment-1673685).  If implemented properly perhaps it could be a really convenient way to report comments and to purge comments.  Also maybe there could be some happy medium where there is some integration between an extension and a desktop GUI/website, where the extension provides extra convenience on top of the website/GUI rather than being standing alone.  This is something that has been implemented by [Simplify Jobs](https://simplify.jobs).

**Comment by Gowixx** [comment]
![image](https://user-images.githubusercontent.com/57415348/154408202-4ddaebe0-5d87-46f4-9fb1-f2ef456a8dfd.png)
Please look at the context and don't repost what others say

**Reply by TechStudent10** [reply]
That was from 15 days ago.

**Reply by Gowixx** [reply]
Same topic

**Comment by CodeDoctorDE** [comment]
PyScript can be used for this. It's very new but is very powerful
